promscale/query/query-traces.md

```diff
@@ -0,0 +1,425 @@
+            and span.status_code = 'error'
+            and span.service_name = 'my_service'
+            and span.span_tags -> 'container_name' = '"my_container"'
+    where 
+            mem.value != 'NaN'
+        and mem.time >= now() - interval '1 week'
+        and span.start_time >= now() - interval '1 week'
+        and mem.container = 'my_container'
+        and mem.instance = 42
+    group by 1
+    order by 1
+```
+There's quite a bit going on in this query, so lets break it down.
+First lets have a look at our metric table and its filters:
+
+1. We want some actual memory usage values so we filter out some misses where the value was reported `NaN`
+2. We explicitly specify last week as the time window to allow planner to eliminate unecessary partitions on the planning stage. This should speed up our query significantly
+3. We explicitly limit the metrics to the container and instance we're interested in. This is also aimed at giving planner more freedom in dealing with the query.
+4. The join clause itself is matching only on the generated `time_bucket`. More on this later.
+
+The `span` view we approach slightly differently:
+
+1. The only qual in the `where` clause we specify is the `start_time` matching that of the metric for the same reason
+2. We specify a number of filters in the join condition instead for several reasons:
+ - We do need to filter out irrelevant rows from the `span` view
+ - We want to keep the entries from the metric table to have a good measure of memory consumption regardless if there were errors or not. If we instead put these quals into the `where` clause, all the rows without errors matching the quals would've been filtered out.
+
+Further we'll see how we can leverage full power of SQL by combining all these primitives together.
+
+## 3.4. Grouping <a name="para-3-4"></a>
+
+One of the strengths of SQL in general and PostgreSQL in particular are aggregate functions which perform various operations on groups. We can group data on any set of columns, includinig fields of `tag_map` columns. Below is a simple example:
+
+```SQL
+select
+        time_bucket('30 minutes', start_time)   as "time",
+        span_tags -> 'pwlen'                    as "pwlen",
+        count(*)                                as "cnt"
+    from span
+    where 
+            start_time >= now() - interval'1 day'
+        and service_name = '${service}'
+    group by 1, 2
+```
+
+Here we use group by both time and the `pwlen` field of the `span_tags` column using their position in the `select` list. `count(*)` in this case is an [aggregate function](https://www.postgresql.org/docs/current/functions-aggregate.html) that counts the number of rows in the group, in our case a 30-minute window with the same value of `pwlen` field.
+
+Now we can build a more complicated query using both techniques. Lets see how error rate correlates with the memory usage:
+
+```SQL
+with joined as
```

Comment:
Have you tested the query to make sure it works?

Reply:
Same here, we'll leave this query for a later point in time.